File: Array&String/HashTable/CPP/how_many_numbers_are_smaller_than_the_current_number.cpp
#include <map>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> smallerNumbersThanCurrent(vector<int>& nums) {
        map<int, int> cnt;
        for (int x: nums) {
            cnt[x]++;
        }
        auto it = cnt.begin();
        int prev = it->second;
        if (it != cnt.end()) ++it;
        cnt.begin()->second = 0;
        for (; it != cnt.end(); ++it) {
            int cur = it->second;
            it->second = prev;
            prev += cur;
        }
        vector<int> res;
        for (int x : nums) res.push_back(cnt[x]);
        return res;
    }
};

File: Array&String/HashTable/how_many_numbers_are_smaller_than_the_current_number.py
from collections import Counter

class Solution:
    def smallerNumbersThanCurrent(self, nums: list[int]) -> list[int]:
        cnt = sorted(list(item) for item in Counter(nums).items())
        prev = cnt[0][1]
        cnt[0][1] = 0
        for i in range(1, len(cnt)):
            cur = cnt[i][1]
            cnt[i][1] = prev
            prev += cur
        cnt = dict(cnt)
        return [cnt[x] for x in nums]
